msticpy/nbtools/timeline.py

```diff
@@ -0,0 +1,940 @@
+    output_notebook()
+
+    height: int = kwargs.pop("height", None)
+    width: int = kwargs.pop("width", 900)
+    ref_time: Any = kwargs.pop("ref_time", None)
+    title: str = kwargs.pop("title", None)
+    time_column: str = kwargs.pop("time_column", "TimeGenerated")
+    source_columns: list = kwargs.pop("source_columns", None)
+    legend_pos: str = kwargs.pop("legend", None)
+    show_yaxis: bool = kwargs.pop("yaxis", False)
+    show_range: bool = kwargs.pop("range_tool", True)
+
+    if not source_columns:
+        source_columns = ["NewProcessName", "EventID", "CommandLine"]
+
+    def_source_columns = set(source_columns) if source_columns else set()
+
+    # Process the input dictionary
+    # Take each item that is passed and fill in blanks and add a y_index
+    tool_tip_columns: Set[str] = set()
+    min_time = pd.Timestamp(pd.Timestamp.max)
+    max_time = pd.Timestamp(pd.Timestamp.min)
+    y_index = 0
+    # Shift the Viridis palatte so we lose the top, harder-to-see colors
+    series_count = len(data)
+    palette_size = min(256, series_count + int(series_count / 2))
+    colors = viridis(palette_size)
+    for ser_name, series_def in data.items():
+        data_columns: Set[str] = set()
+        series_data = series_def["data"]
+
+        # if the series has source columns, use those
+        src_cols = series_def.get("source_columns", def_source_columns)
+        data_columns.update(src_cols if src_cols else def_source_columns)
+
+        # add these columns to the tool tip column set
+        tool_tip_columns.update(data_columns)
+
+        time_col = series_def.get("time_column", None)
+        if not time_col:
+            time_col = time_column
+            series_def["time_column"] = time_col
+
+        min_time = min(min_time, series_data[time_col].min())
+        max_time = max(max_time, series_data[time_col].max())
+        data_columns.update([time_col])
+        # Create the Column data source to plot
+        graph_df = series_data[list(data_columns)].copy()
+        graph_df["y_index"] = y_index
+        series_def["y_index"] = y_index
+        ser_color = series_def.get("color", None)
+        if not ser_color:
+            ser_color = colors[y_index % palette_size]
+            series_def["color"] = ser_color
+        # Wrap tooltip lines longer than 50 chars
+        _wrap_df_columns(graph_df, 50)
+        series_def["source"] = ColumnDataSource(graph_df)
+        y_index += 1
+
+    # build the tool tips from all specified columns
+    tool_tip_items = [(f"{col}", f"@{col}") for col in tool_tip_columns]
+    hover = HoverTool(tooltips=tool_tip_items, formatters={"Tooltip": "printf"})
+
+    title = f"Timeline: {title}" if title else "Event Timeline"
+    start_range = min_time - ((max_time - min_time) * 0.1)
+    end_range = max_time + ((max_time - min_time) * 0.1)
+    height = height if height else _calc_auto_plot_height(len(data))
+    plot = figure(
+        x_range=(start_range, end_range),
```

Comment:
for the multi data set plots we probably want to set the y_range as well. Something like y_range=(0, len(data))

Reply:
Agreed - +/- 25% of min/max